### PR TITLE
Open multiple articles in browser

### PR DESCRIPTION
**Pull request by tkerdonc**
This adds commands to open multiple unread articles from the same
feed in the browser. These commands are named
open-all-in-browser and open-all-in-browser-and-mark-read with the
difference of marking opened items as read or not. Both commands are
available from the feed list view and the article list view.

To avoid mishaps, the number of articles that can be opened by these
commands is limited by the max-browser-tabs configuration command.

### CODE REVIEW

File: src/itemlist_formaction.cpp [hunk 1]
```diff
@@ -109,6 +109,38 @@ void itemlist_formaction::process_operation(operation op, bool automatic, std::v
...
+			if (tabcount < v->get_cfg()->get_configvalue_as_int("max-browser-tabs")) {
+				if (item->unread()){
+					v->open_in_browser(item->link());
+					item->set_unread(false);
+					tabcount += 1;
+				}
+			} else {
+				break;
+			}
+		}
```
Comment (Minoru) [inline src/itemlist_formaction.cpp]: The code here is pretty much the same as in feedlist. Let's scrap the boilerplate, eh? (I realise the codebase doesn't yet look like we care, but we do. I'm gradually cleaning it up. You can help me by writing this one bit cleanly in the first place.)

You can create `include/formactions.h` and `src/formactions.cpp` and put this code there. Fit it into one function that accepts a feed and a boolean indicating if you should mark articles read. The logging code is better kept in `*_formaction.cpp` files as it's easier to grep the code then.

**Review by Minoru — CHANGES_REQUESTED** [review]
A few more things:
- please update `doc/example-config` to include `max-browser-tabs` setting;
- can you write some tests? Setting `browser` to `echo` and writing into predefined temporary file—`/tmp/newsbeuter.txt` should do—that you delete later should be enough to make it all testeable. I'm willing to answer questions if you are willing to work on that ;)

File: src/keymap.cpp [hunk 2]
```diff
@@ -35,6 +35,8 @@ static op_desc opdescs[] = {
 	{ OP_RANDOMUNREAD,		"random-unread",			"^K",	_("Go to a random unread article"),		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE },
 	{ OP_OPENBROWSER_AND_MARK, "open-in-browser-and-mark-read",                   "O",    _("Open article in browser and mark read"),           KM_ARTICLELIST },
 	{ OP_OPENINBROWSER,		"open-in-browser",			"o",	_("Open article in browser"),			KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE },
+	{ OP_OPENALLINBROWSER, "open-all-in-browser",                   "NOKEY",    _("Open all unread items of selected feed in browser"),          KM_FEEDLIST | KM_ARTICLELIST },
```
Comment (Minoru) [inline src/keymap.cpp]: Instead of `"NOKEY"`, specify an empty string. This does the right thing in terms of listing the operations in the help dialog as "unbound".

**Comment by tkerdonc** [comment]
Thanks for the detailed review, I'll work on these points.

**Comment by Minoru** [comment]
Hope I didn't overwhelm you ._.

Regarding tests: I didn't think it through, `echo` won't be enough to write to a file. Instead, you'll probably have to create a little shell script that writes to a file.

**Comment by tkerdonc** [comment]
No worries Minoru, on the contrary, this gives me a clear todo list.
Regarding tests, setting browser to "echo %u >> /tmp/newsbeuter.txt" seems to do the trick.

**Comment by tkerdonc** [comment]
Committed changed regarding the aforementioned points, except for testing. I might be missing something obvious here, but I could not find an non intrusive way to automatically trigger operations such as OP_OPENALLUNREADINBROWSER (the process_operation being (rightfully) private).
Is there an stfl function that could help me here ? Another thought is that this might be doable by implementing a test_view class inherinting from view...

Ideas ?

File: doc/example-config [hunk 3]
```diff
@@ -612,6 +612,18 @@
 #
 # mark-as-read-on-hover yes
 
+####  max-browser-tabs
+#
+# Sets the maximum number of article that can be opened by the open in the
```
Comment (Minoru) [inline doc/example-config]: "Article*s*", not "article". I prefer to copy and paste the descriptions from configcommands.dsv—it's simply easier.

File: include/formaction.h [hunk 4]
```diff
@@ -72,6 +72,7 @@ class formaction {
 
 
 		void start_bookmark_qna(const std::string& default_title, const std::string& default_url, const std::string& default_desc, const std::string& default_feed_title);
+		void open_unread_items_in_browser(std::shared_ptr<rss_feed> feed , bool markread);
```
Comment (Minoru) [inline include/formaction.h]: That's not what I had in mind, but I admit that I just do "high-level" thinking and leave details for you to sort out.

I asked to put this into `formaction*s*`, a new, separate file, specifically because I didn't want to put it in `formaction` class. `formaction` is for sutff shared across all kinds of forms, while this "open all tabs" business is specific to feed- and itemlist.

Of course, I overlooked the fact that the method needs to know maximum number of tabs to open, thus needs access to `configcontainer` object. Passing `cfg` is tedious, so the method has to go into a class that inherits from `view`.

So here's another idea (which you are welcome to argue with if you see any drawbacks in it): create a mixin class that derives from `formaction`, put the method there, then derive `feedlist_formaction` and `itemlist_formaction` from that mixin (virtually, because diamond problem). What do you think?

File: src/keymap.cpp [hunk 5]
```diff
@@ -35,6 +35,16 @@ static op_desc opdescs[] = {
 	{ OP_RANDOMUNREAD,		"random-unread",			"^K",	_("Go to a random unread article"),		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE },
 	{ OP_OPENBROWSER_AND_MARK, "open-in-browser-and-mark-read",                   "O",    _("Open article in browser and mark read"),           KM_ARTICLELIST },
 	{ OP_OPENINBROWSER,		"open-in-browser",			"o",	_("Open article in browser"),			KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE },
+    {
```
Comment (Minoru) [inline src/keymap.cpp]: From a history standpoint, I'd rather keep this change to one commit. I mean, preserve the formatting as is, and let us change it all at once in #383.

Also, we use tabs for indentation.

**Review by Minoru — APPROVED** [review]

**Comment by Minoru** [comment]
Hmm, I should've submitted my latest comments as a separate review, not just bunch of messages. Oh well.

> non intrusive way to automatically trigger operations such as OP_OPENALLUNREADINBROWSER (the process_operation function being (rightfully) private)

You can call `process_op`, which is public; it'll fall through the cases and call `process_operation` in `default` clause.

**Comment by tkerdonc** [comment]
Thanks for the information, I should be able to implement these changes next week.

**Comment by Minoru** [comment]
Okay, will look forward to that. Have a nice weekend!

**Comment by tkerdonc** [comment]
Hi again. Just committed the latest changes, featuring the list_formaction intermediary class, and some tests. Two point I would like to discuss though

- Might have missed the diamond problem, the way I implemented thing was to have itemlist_formaction and feedlist_formaction derive from list_formaction, wich itself derives from formaction. Sa maybe not what you had in mind, but seems to do the trick.
- I only added a test running on the itemlist  formaction, I might need some advice regarding testing from the feedlist formaction. It uses the feedposname variable in order to identify the selected feed, wich I can't seem to access from my test case. It is proper to stfl right ?

And you'll notice I went lazy and only tested my two new commands, while all the other ones could be tested here. This is something I could happily work on if you want, but better save it for another PR no ?

Looking forward to your review.

File: src/list_formaction.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace newsbeuter {
+
+list_formaction::list_formaction(view * v, std::string formstr)
+:formaction(v, formstr)
+{}
+
+void list_formaction::open_unread_items_in_browser(std::shared_ptr<rss_feed> feed , bool markread){
+    int tabcount = 0;
+    for (auto item : feed->items()) {
```
Comment (tsipinakis) [inline src/list_formaction.cpp]: This is also indented with spaces.

File: test/itemlist_formaction.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+	v.set_config_container(cfg);
+	c.set_view(&v);
+
+	std::shared_ptr<itemlist_formaction> itemlist(new itemlist_formaction(&v, itemlist_str));
+	itemlist->set_feed(feed);
+
+	SECTION("Itemlist Formaction"){
+	
+		SECTION("Open all unread in browser, unread > max-browser-tabs"){
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: Judging by the code, it's `unread == max-browser-tabs`. I think it's okay to update this to `>=` and let it be.

File: test/itemlist_formaction.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+			openedItemsTarget = maxItems;
+			itemlist->process_op(OP_OPENALLUNREADINBROWSER);
+
+			std::ifstream myfile (test_browser_file);
+			openedItemsCount = 0;
+  			if (myfile.is_open())
+  			{
+    				for (int i = 0; i < maxItems; i ++)
+    				{
+					if ( std::getline (myfile,line) )
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: Why isn't this indented further?

File: test/itemlist_formaction.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+						REQUIRE(line == test_url + std::to_string(i));
+						openedItemsCount += 1;
+					}
+    				}
+    				myfile.close();
+			}
+			REQUIRE(openedItemsCount == openedItemsTarget);
+			std::remove(test_browser_file.c_str());
+		}
+
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: I won't repeat the comments for the previous two test cases for the last two, it's all the same stuff.

File: test/itemlist_formaction.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	std::string line;
+	int itemCount = 6;
+	int openedItemsCount = 0;
+	int lineCount = 0;
+	int maxItems = 4;
+	int openedItemsTarget = 0;
+
+	configcontainer * cfg = new configcontainer();
+	cfg->set_configvalue("browser", "echo %u >> " + test_browser_file);
+	cfg->set_configvalue("max-browser-tabs", std::to_string(maxItems));
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: This line belongs to the first test case, because the rest sets `max-browser-tabs` themselves.

File: test/itemlist_formaction.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	newsbeuter::view v(&c);
+	std::string test_browser_file = "/tmp/test_browser.txt";
+	std::string test_url = "http://test_url";
+	std::string cache_path = "test-cache.db";
+	std::string line;
+	int itemCount = 6;
+	int openedItemsCount = 0;
+	int lineCount = 0;
+	int maxItems = 4;
+	int openedItemsTarget = 0;
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: I have two comments regarding all of these `int`s:
1. Some of them look like constants, but I'm not entirely sure. If they are, just mark them `const`.
2. It's kinda hard to read the tests because I have to flick between initializations here and the test cases. I suggest to violate DRY principle and re-declare all of these (save for constants, if any) inside test cases.

**Review by Minoru — CHANGES_REQUESTED** [review]
The code's fine. Tests still require some polishing.

File: test/itemlist_formaction.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+	SECTION("Itemlist Formaction"){
+	
+		SECTION("Open all unread in browser, unread > max-browser-tabs"){
+			openedItemsTarget = maxItems;
+			itemlist->process_op(OP_OPENALLUNREADINBROWSER);
+
+			std::ifstream myfile (test_browser_file);
+			openedItemsCount = 0;
+  			if (myfile.is_open())
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: Good catch! But how did you do it? Is there some button or setting in GitHub UI that lets you see tabs, or do you just check the code out locally?

Comment (tsipinakis) [inline test/itemlist_formaction.cpp]: I checked out the code locally, apparently `git diff` highlights spaces followed by a tab character.

**Comment by Minoru** [comment]
> […] itemlist_formaction and feedlist_formaction derive from list_formaction, wich itself derives from formaction

That'll do. We can always change it later if it begins to constrain us in any way. Thanks!

> I might need some advice regarding testing from the feedlist formaction. It uses the feedposname variable in order to identify the selected feed, wich I can't seem to access from my test case.

I don't think you even need to access it. I'm not 100% sure, but I believe the position is set to the first feed when you create the form. So if you create just one feed, put it into `std::vector<…>` and use `set_feedlist`, you'll automatically have feedpos pointing to this feed. From there, testing shouldn't be any different from itemlist.

If this fails, do tell me and I'll look into it in more detail (writing off the top of my head ATM.)

> It is proper to stfl right ?

I don't think I understood that, sorry. Can you rephrase?

> And you'll notice I went lazy and only tested my two new commands, while all the other ones could be tested here. This is something I could happily work on if you want, but better save it for another PR no ?

You mean the rest of operations that forms support? That'll be fantastic! (I spent last week writing tests for cache module and don't think I want to write any more anytime soon unless I absolutely need to :)

And indeed, this is better reserved for a separate pull request.

**Comment by tkerdonc** [comment]
Thanks for all the remarks guys, I believe the missing tabs were caused by moving the code around from its previous class. Will rework on all the aforementioned points.

> > It is proper to stfl right ?
> I don't think I understood that, sorry. Can you rephrase?
My bad, what I meant was that this _feedposname_ variable is only read in feedlist_formaction.cpp and defined in headers and .stfl files of the stfl folder. So my guess is that this value is only set by direct user interaction. Unless I am missing a critical link between _feedpos_ (who is properly set to the feed position) and _feedposname_. I'll dig a bit further and get back to you, would you like me to push a failing feedlist_formaction test class reproducing this problem ?

**Comment by Minoru** [comment]
Ah, I see what you mean.

> So my guess is that this value [feedposname] is only set by direct user interaction.

Normally, yes. But:

1. I'm pretty sure there's some default, probably zero. It doesn't look like we set `feedposname` ourselves anywhere in the code.
2. While digging into this, I found that you can actually set `feedposname` from the code like this `f->set("feedposname", 3)` (where `f` is an object of type `stfl::form`. ) I didn't test this, I just found it in `include/stflpp.h`.

> would you like me to push a failing feedlist_formaction test class reproducing this problem ?

Yes, please do!

File: test/itemlist_formaction.cpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+using namespace newsbeuter;
+
+TEST_CASE("itemlist_formaction::process_op()") {
+
+	controller c;
+	newsbeuter::view v(&c);
+	std::string test_browser_file = "/tmp/test_browser.txt";
+	std::string test_url = "http://test_url";
+	std::string cache_path = "test-cache.db";
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: Sorry, it's only now that I realized that you don't even need a file—you can just use in-memory DB by passing `":memory:"` instead of the path. (You'll still need `TempFile` for `test_browser_file` though.)

**Comment by Minoru** [comment]
Um, nope. You did a merge, not a rebase. Should I step in and fix it? You can follow [this guide](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) to give me push access to your branch.

**Comment by tkerdonc** [comment]
Hi Minoru, good thing you are around ^^. You'll probably have an easier time than I fixing the mess up, thanks. Edits from maintainers are enabled.

**Comment by Minoru** [comment]
Done! Looks like you failed at the very last step: after rebasing onto master and squashing your commits into one, you should've done `git push --force`, but you probably did this instead:

* `git push`, which complained
* `git pull`, which resulted in a merge (with one conflict)
* resolve conflict, `git add`, `git commit`
* finally, `git push` again, this time without any problems

Can't blame you, force-pushes and rebasing are quite complicated and dangerous things…

**Comment by Minoru** [comment]
But I don't see the test for `feedlist`, and it looks like you haven't addressed any of my review comments yet. Is that correct, or did I miss something? (I still have a clone with all your changes, so I can still recover everything if I messed up.)

**Comment by tkerdonc** [comment]
The steps you mentionned above definitely sound familiar, I need to practice git more. Thanks for stepping in. And yeah, the changes I was attempting to commit seems to have been lost, but I still have a local copy, I'll do a fresh checkout, overwrite and push back to the branch, then comment here.

**Comment by tkerdonc** [comment]
Pushed the last changes, with a refactoring of the itemlist_formaction test class.
The start of the feedlist formaction is also attached but does not test anything yet. I still plan on investigating that, but here is the current status, in case I am missing something obvious. As you mentionned, set("feedposname", "0") does the trick, but in order for this to work, I need to involve the controller, and have its "feeds" attribute initialized, which seems to only be done by the url_reader.
I'll keep looking, but if you have any pointers, they would (again ^^) be very appreciated.

**Review by Minoru — CHANGES_REQUESTED** [review]

**Comment by Minoru** [comment]
Looked into it for a bit, but no luck yet. I'll have a hectic schedule this weekend and Monday, so don't know if I will be able to crack it, but I'll keep trying as time permits.

**Comment by Minoru** [comment]
To be clear: this will definitely be merged. You've done the coding, refactoring, documentation and half the tests. The other half we can just test by hand for now. It's none of your fault that I can't make heads or tails out of project I'm responsible for ._.

Polish the tests you already have, and if neither of us can figure out how to test `feedlist` automatically, I'll just merge this as-is.

**Comment by tkerdonc** [comment]
Allright, thanks again for the patience, review and detailed instructions. I'll be busy too this weekend, but should be able to integrate these on monday. See you then !

**Comment by Minoru** [comment]
JFTR, I gave up on trying to make `formaction` test work. Looks like it's necessary to have STFL actually draw something for `feedposnaame` property to be initialized.

Have a nice weekend!

File: test/itemlist_formaction.cpp [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+			std::ifstream browserFileStream (browserfile.getPath());
+			openedItemsCount = 0;
+			if (browserFileStream.is_open()) {
+				for (int i = 0; i < maxItemsToOpen; i ++) {
+					if ( std::getline (browserFileStream,line) ) {
+						REQUIRE(line == test_url + std::to_string(i));
+						openedItemsCount += 1;
+					}
+				}
```
Comment (Minoru) [inline test/itemlist_formaction.cpp]: I just realized that writing the check this way will make tests pass when they shouldn't.

Imagine that we broke our code somehow and now we open items twice. We do one pass over items and then another one. `browserfile` will look like this (for three items):
```
url1
url2
url3
url1
url2
url3
```

The test will only examine the first three lines, and won't see that we broke the code.

Rewriting this to `while(getline(…))` or something should fix the problem.

Comment (Minoru) [inline test/itemlist_formaction.cpp]: There's another problem with this check, namely it's dependence on items order. If we were to start opening items randomly, or sorted by date or something, this check will start failing with errors like `"http://test_url3" != "http://test_url5"`, which isn't helpful at all (and tests are supposed to make errors obvious, not just tell of their existence.)

To fix this, you'll probably have to build an `unordered_set` of test URLs (while you're adding items to feed) and then test if the line you read from the file is present in the set (and then remove it to avoid duplication problem described in previous comment). Also add [an `INFO` macro](https://github.com/philsquared/Catch/blob/master/docs/logging.md#streaming-macros) right before `REQUIRE` to explain what the check means—this should make the error less cryptic.

(Misclicked the button, so this comment is now going to be a separate review. Oh well.)

**Comment by tkerdonc** [comment]
Integrated you latest remarks and removed the feedlist test (no luck on my end either, testing this may require to modify newsbeuter's source). Let me know what you think.

**Comment by Minoru** [comment]
I think that this is ready to be merged. Thank you for working on this!

**Comment by Minoru** [comment]
Oh, and I'm going to squash your commits into one because there's no reason to keep this history.

**Comment by Minoru** [comment]
Oh, and regarding that "more tests for `itemlist_formaction`" thing: feel free to open issues if you stumble upon anything while working on this, be it a question, or a bug (these are better fixed *after* we have tests anyway), or just anything at all.

**Comment by tkerdonc** [comment]
Excellent, thanks a lot for the support all along. I'll now have a go at the itemlist_formaction tests then, and keep you updated.
